dispatch/dispatch.go

```diff
@@ -493,29 +491,30 @@ func (ag *aggrGroup) empty() bool {
 }
 
 // flush sends notifications for all new alerts.
-func (ag *aggrGroup) flush(notify func(...*types.Alert) bool) {
+func (ag *aggrGroup) flush(ctx context.Context, nf notifyFunc) {
 	if ag.empty() {
 		return
 	}
 
+	now, ok := notify.Now(ctx)
+	if !ok {
+		level.Error(ag.logger).Log("msg", "now missing")
+		return
+	}
+
 	var (
 		alerts      = ag.alerts.List()
 		alertsSlice = make(types.AlertSlice, 0, len(alerts))
-		now         = time.Now()
 	)
 	for _, alert := range alerts {
 		a := *alert
```

Comment:
I suspect we can remove this, but will do so in another PR.

Reply:
If I understand the intent correctly this is passing the reference time through the context and every notifier should use the `.*At(now time.Time)` variant. As we don't (plan) to modify the alert we don't need to take a deep copy (some notifiers call types.Alerts).

Dropping this copy might be hard. On line 528 (529) the UpdatedAt time is compared between the copy and the storage of the aggregation group. Without a copy these numbers will always be the same.

I am wondering if there are ways to make sure that integrations never call the `Resolved()`, `Status()`? Given that alertmanager is likely to see an increase in number of integrations that appears desirable to me to ease writing and reviewing these.

Given that the model.Alert is copied and one needs the reference time. What is your thought on creating something like a model.AlertsSnapshot that holds an immutable snapshot of the alert and the status/reference time? This would ease writing and reviewing the notifiers, protect the core from accidental modificiations?

Reply:
> Dropping this copy might be hard. On line 528 (529) the UpdatedAt time is compared between the copy and the storage of the aggregation group. Without a copy these numbers will always be the same.

Oh well spotted!

> What is your thought on creating something like a model.AlertsSnapshot that holds an immutable snapshot of the alert and the status/reference time?

This is what I tried at first, but I found it to be quite an intrusive change (https://github.com/prometheus/alertmanager/issues/3351#issuecomment-1567365253). I haven't ruled it out though - do you think that would be a better way to fix this?

Reply:
I think longterm we are better off with a genuine snapshot. It eases writing integrations, it eases code review of them. I think some of the type changes we can probably automate with gofmt?

Reply:
Something like this?

```go
type AlertSnapshot struct {
	Alert
	// status of the alert at the time it was snapshot.
	status model.AlertStatus
}

func (a *AlertSnapshot) Resolved() bool {
	return a.status == model.AlertResolved
}

func (a *AlertSnapshot) Status() model.AlertStatus {
	return a.status
}

// Need to override String() as otherwise it will call Resolved()
// of model.Alert not AlertSnapshot.
func (a *AlertSnapshot) String() string {
	s := fmt.Sprintf("%s[%s]", a.Name(), a.Fingerprint().String()[:7])
	if a.Resolved() {
		return s + "[resolved]"
	}
	return s + "[active]"
}
```

Reply:
> UpdatedAt, Name, Status, Resolved, Labels, Annotations, StartsAt, EndsAt, GeneratorURL, Fingerprint, ResolvedAt

We would all inherit these from `model.Alert` embedded in the struct?

Reply:
Yeah but from my point of view the trade-off is whether we want all of these fields to be mutable.

Reply:
Oh – I think I misunderstood. By mutable/immutable are you talking about the fields being package-public (capitalized) or package-private (starting with lowercase)?

Reply:
Yes. Starting with lower case. I wanted to see how an implementation of this looks like and got the below to compile (and pass most tests but it requires some clean ups).

For me the biggest benefit is that it creates an immutable snapshot and simplifies API. The notification handlers are unable to modify the underlying Alert and the API removes the `*At(time.Time)` functions.


Please see:
https://github.com/prometheus/alertmanager/compare/main...zecke:alertmanager:zecke-alternative-ends-at

Reply:
Sure. Let me fix the formatting and remove some of the repeated code in the tests.